backend/de.metas.contracts/src/main/java/de/metas/contracts/finalinvoice/invoicecandidate/FlatrateTermModular_Handler.java

```diff
@@ -0,0 +1,312 @@
+				.orElseThrow(() -> new AdempiereException("Product cannot be missing at this point !")
+						.appendParametersToMessage()
+						.setParameter("C_Flatrate_Term_ID", term.getC_Flatrate_Term_ID()));
+	}
+
+	@NonNull
+	private I_C_Invoice_Candidate createCandidateFor(
+			@NonNull final I_C_Flatrate_Term modularContract,
+			@NonNull final ModuleConfig moduleConfig,
+			@NonNull final YearAndCalendarId yearAndCalendarId,
+			@Nullable final PricingSystemId pricingSystemId)
+	{
+		final I_C_Invoice_Candidate invoiceCandidate = InterfaceWrapperHelper.newInstance(I_C_Invoice_Candidate.class);
+
+		invoiceCandidate.setAD_Table_ID(InterfaceWrapperHelper.getTableId(I_C_Flatrate_Term.class));
+		invoiceCandidate.setRecord_ID(modularContract.getC_Flatrate_Term_ID());
+		invoiceCandidate.setAD_Org_ID(modularContract.getAD_Org_ID());
+		invoiceCandidate.setC_Flatrate_Term_ID(modularContract.getC_Flatrate_Term_ID());
+		invoiceCandidate.setC_Async_Batch_ID(modularContract.getC_Async_Batch_ID());
+
+		InvoiceCandidateLocationAdapterFactory
+				.billLocationAdapter(invoiceCandidate)
+				.setFrom(ContractLocationHelper.extractBillLocation(modularContract));
+
+		invoiceCandidate.setIsTaxIncluded(modularContract.isTaxIncluded());
+
+		//
+		invoiceCandidate.setLine(moduleConfig.getSeqNo().toInt());
+		invoiceCandidate.setM_Product_ID(moduleConfig.getProductId().getRepoId());
+		invoiceCandidate.setInvoicingGroup(moduleConfig.getInvoicingGroup());
+		invoiceCandidate.setProductName(moduleConfig.getName());
+		//
+		invoiceCandidate.setHarvesting_Year_ID(yearAndCalendarId.yearId().getRepoId());
+		invoiceCandidate.setC_Harvesting_Calendar_ID(yearAndCalendarId.calendarId().getRepoId());
+		//
+		invoiceCandidate.setM_PricingSystem_ID(PricingSystemId.toRepoId(pricingSystemId));
+
+		return invoiceCandidate;
+	}
+
+	@NonNull
+	private TaxId getTaxId(
+			@NonNull final I_C_Invoice_Candidate invoiceCandidate,
+			@NonNull final TaxCategoryId taxCategoryId)
+	{
+		final BPartnerLocationAndCaptureId bPartnerLocationAndCaptureId = getBpartnerLocationAndCapture(invoiceCandidate);
+		final OrgId orgId = OrgId.ofRepoId(invoiceCandidate.getAD_Org_ID());
+		final SOTrx soTrx = SOTrx.ofBooleanNotNull(invoiceCandidate.isSOTrx());
+
+		final Tax tax = taxDAO.getBy(TaxQuery.builder()
+											 .orgId(orgId)
+											 .bPartnerLocationId(bPartnerLocationAndCaptureId)
+											 .dateOfInterest(invoiceCandidate.getDeliveryDate())
+											 .soTrx(soTrx)
+											 .taxCategoryId(taxCategoryId)
+											 .build());
+
+		if (tax == null)
+		{
+			throw TaxNotFoundException.builder()
+					.taxCategoryId(taxCategoryId)
+					.isSOTrx(soTrx.toBoolean())
+					.billDate(invoiceCandidate.getDeliveryDate())
+					.orgId(orgId)
+					.build();
+		}
+
+		return tax.getTaxId();
+	}
+
+	@NonNull
+	private BPartnerLocationAndCaptureId getBpartnerLocationAndCapture(@NonNull final I_C_Invoice_Candidate invoiceCandidate)
```

Comment:
why not using de.metas.invoicecandidate.api.impl.InvoiceCandBL#getBillLocationId?

Reply:
will do